app/src/org/commcare/dalvik/application/CommCareApplication.java

```diff
@@ -377,6 +424,96 @@ private int initializeAppResources(CommCareApp app) {
         }
     }
 
+    /**
+     * @return all installed ApplicationRecords, in alphabetical order
+     */
+    public ArrayList<ApplicationRecord> getInstalledAppRecords() {
+        ArrayList<ApplicationRecord> records = new ArrayList<>();
+        for (ApplicationRecord r : getGlobalStorage(ApplicationRecord.class)) {
+            records.add(r);
+        }
+        Collections.sort(records, new Comparator<ApplicationRecord>() {
+
+            @Override
+            public int compare(ApplicationRecord lhs, ApplicationRecord rhs) {
+                return lhs.getDisplayName().compareTo(rhs.getDisplayName());
+            }
+
+        });
+        return records;
+    }
+
+    /**
+     * @return all ApplicationRecords that are installed and NOT archived
+     */
+    public ArrayList<ApplicationRecord> getVisibleAppRecords() {
+        ArrayList<ApplicationRecord> visible = new ArrayList<>();
+        for (ApplicationRecord r : getInstalledAppRecords()) {
+            if (!r.isArchived()) {
+                visible.add(r);
+            }
+        }
+        return visible;
+    }
+
+    /**
+     * @return  all ApplicationRecords that are installed AND are not archived AND have MM verified
+     */
+    public ArrayList<ApplicationRecord> getReadyAppRecords() {
+        ArrayList<ApplicationRecord> ready = new ArrayList<>();
+        for (ApplicationRecord r : getVisibleAppRecords()) {
+            if (r.resourcesValidated()) {
+                ready.add(r);
+            }
+        }
+        return ready;
+    }
+
+    public boolean visibleAppsPresent() {
+        return getVisibleAppRecords().size() > 0;
+    }
+
+    /**
+     * @return the list of all installed apps as an array
+     */
+    public ApplicationRecord[] appRecordArray() {
+        ArrayList<ApplicationRecord> appList = CommCareApplication._().getInstalledAppRecords();
+        ApplicationRecord[] appArray = new ApplicationRecord[appList.size()];
+        int index = 0;
+        for (ApplicationRecord r : appList) {
+            appArray[index++] = r;
+        }
+        return appArray;
+    }
+
+    /**
+     * @param index an index into the global table of installed apps
+     * @return the ApplicationRecord that corresponds to the given index in the global list of
+     * installed apps
+     */
+    public ApplicationRecord getAppAtIndex(int index) {

```

Comment:
Sorry, I'm not totally following the problem, or potential solutions. This method is being used by SingleAppManagerActivity to know which app it is supposed to be displaying, based upon the position of the list item that was clicked in AppManagerActivity to get there. Since getInstalledAppRecords() always returns the apps in alphabetical order, and the list view in AppManagerActivity is populated by the appRecordArray generated from that, the ordering should always be the same as far as I'm seeing. It would probably be more proper to implement getAppAtIndex() using appRecordArray() for consistency's sake, but that isn't really a substantive change. What's your concern for the inconsistency?